### ISSUE TITLE
[postgres] Parse error for unnest with ordinality

### ISSUE THREAD

**Issue by mfvanek** [sqlfluff/sqlfluff#3843]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Parser fails on `unnest` queries `with ordinality`.
Example:
```sql
select * from unnest(array['123', '456']) with ordinality as a(val, row_num)
```
throws
```
SQLParseError("Line 1, Position 64: Found unparsable section: 'with ordinality as a(val, row_num)'")
```

### Expected Behaviour

` with ordinality` is a valid syntax

> If the WITH ORDINALITY clause is specified, an additional column of type bigint will be added to the function result columns. 

https://www.postgresql.org/docs/current/queries-table-expressions.html

### Observed Behaviour

`SQLParseError("Line 1, Position 64: Found unparsable section: 'with ordinality as a(val, row_num)'")`

It looks similar to https://github.com/sqlfluff/sqlfluff/issues/1329 and https://github.com/sqlfluff/sqlfluff/issues/2787

### How to reproduce

Run parser on
`select * from unnest(array['123', '456']) with ordinality as a(val, row_num)`
or
```python
@pytest.mark.parametrize(
    "raw",
    [
        "select * from unnest(array['123', '456', '789', '258', '147']) with ordinality as a(val, row_num)",
        "select c.conrelid::regclass as table_name from pg_catalog.pg_constraint c join lateral unnest(c.conkey) with ordinality as u(attnum, attposition) on true"
    ],
)
def test_unnest_with_ordinality_is_supported(raw: str) -> None:
    cfg = FluffConfig(
        configs={"core": {"exclude_rules": "L009,L011,L017,L025,L027,L028,L031,L044", "dialect": "postgres"}}
    )
    lnt = Linter(config=cfg)
    result = lnt.lint_string(raw)
    assert result.num_violations() == 0
    result = lnt.parse_string(raw)
    assert len(result.violations) == 0
```

### Dialect

postgres

### Version

1.2.1

### Configuration

```
[sqlfluff]
dialect = postgres
exclude_rules = L009,L011,L017,L025,L027,L028,L031,L044

[sqlfluff:rules]
max_line_length = 180

[sqlfluff:rules:L007]
operator_new_lines = before

[sqlfluff:rules:L029]
ignore_words = columns,toast
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by mfvanek** [comment]
@tunetheweb could you please take a look at this issue?

**Comment by Micaelamelo** [comment]
Hello, what's the status of this issue? We are facing this problem with my team, thanks!
